A33-Fault-Injection.md

```diff
@@ -0,0 +1,268 @@
+
+There are currently four ways to set the filter configuration, ordered by ascending priority:
+
+Defining filter config in `http_filters` of HTTP Connection Manager;
+Overriding filter config with `typed_per_filter_config` of VirtualHost from RDS;
+Overriding filter config with `typed_per_filter_config` of Route from RDS;
+Overriding filter config with `typed_per_filter_config` of ClusterWeight from RDS;
+
+Also, there is the upcoming Filter Config Discovery Service (FCDS) (see [issue](https://github.com/envoyproxy/envoy/issues/7867)). It allows application to stream / delta / fetch filter configs. It’s not yet released and lacks documentation. From its source code, it updates the global `http_filters`. So, its priority should be between item 1 and item 2.
+
+### Fault Injection in Envoy xDS API
+
+Fault injection in xDS API is an HTTPFilter named `HTTPFault`. As mentioned above, the fault injection config can be provided by **LDS** (HTTP Connection Manager) and **RDS** (VirtualHost, Route, ClusterWeight). Besides, the fault injection filter offers [global Runtime settings](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/fault_filter#runtime) that can override global default fault injection filter config. For the explanation of Envoy Runtime, see section below.
+
+Currently, the Traffic Director (TD) isn’t forwarding HTTPFilter configs in HTTP Connection Manager from LDS, and there isn’t an exposed API allowing users to directly config fault injection for all HTTP connections. To pave gRPC’s way for more extensive control plane compatibility, we need to support fetching filter config from HTTP Connection Manager among LDS responses.
+
+The Envoy Runtime isn’t yet supported by TD or gRPC. The Runtime path also has a lower priority to implement.
+
+On the other hand, TD has a released support for fault injection via route configuration for HTTP. So, **this design doc focuses on getting fault injection filter config from LDS and RDS**.
+
+### Fault Injection Features
+
+Fault Injection Options
+The full definition of Envoy v3 fault injection proto can be found [here](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/fault/v3/fault.proto#extensions-filters-http-fault-v3-httpfault). In short, it allows users to configure following options:
+
+1. Fixed delay before forwarding the request;
+1. Overridden gRPC status code;
+1. Traffic matcher (upstream cluster name, downstream cluster name, and headers);
```

Comment:
There isn't any detail in here as to what is covered by this part of the feature.

In the envoy docs, I see:

  "upstream_cluster": "...",
  "headers": [],
  "downstream_nodes": [],

`upstream_cluster` makes sense - If set, we'd match on the cluster chosen by the config selector.

`headers` seems straightforward based on the envoy docs: "A match will happen if all the headers in the config are present in the request with the same values (or based on presence if the value field is not in the config)."

`downstream_nodes`, though, I'm not sure what to do with.  Is this what you meant by "downstream cluster name"?  For this it sounds like we would just match the value(s? - what if there are multiple) of the header "x-envoy-downstream-service-node" with the settings in "downstream_nodes" in the fault injection config.

But where does this header come from?  It looks like it's set by envoy, not the user.  Based on "the value is taken from the `--service-node` option" [[ref]](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#x-envoy-downstream-service-node).  We aren't setting this today as far as I'm aware, so does it make sense to support this feature?

Reply:
For the traffic matcher, I updated the design doc that I think that only header-matching is common enough. Both the upstream cluster or downstream nodes features will postponed until users are asking for them.

The headers (or our metadata) can be set either by upstream Envoy or the application. I'm not sure how much will that make.

Reply:
> The headers (or our metadata) can be set either by upstream Envoy or the application.

The ones for the header matchers may be set by the application.  But the one for the "downstream node" feature (which I now see we aren't supporting here) seems to be set only by Envoy, per the ref above.  I'm not clear on what else this may be used for, and I am wondering if setting it is something we need to do in gRPC (or will need to do soon for other reasons).

Reply:
To support this feature, I think we will need to add this feature as well. Technically, gRPC's XdsClient can add `x-envoy-downstream-service-node` or `service-node` header with it's Node info for all processed requests?

But, for this proposal, I hope to focus on the most common usages of the fault injection filter, which is delay, abort, match header, and max active faults.